api/database_engine.go

```diff
@@ -67,7 +67,147 @@ func (e *EverestServer) UpdateDatabaseEngine(ctx echo.Context, namespace, name s
 	return e.proxyKubernetes(ctx, namespace, databaseEngineKind, name)
 }
 
+// GetUpgradePlan gets the upgrade plan for the given namespace.
+func (e *EverestServer) GetUpgradePlan(
+	c echo.Context,
+	namespace string,
+) error {
+	ctx := c.Request().Context()
+	result, err := e.getUpgradePlan(ctx, namespace)
+	if err != nil {
+		return err
+	}
+
+	// No upgrades available, so we will check if our clusters are ready for current version.
+	if len(pointer.Get(result.Upgrades)) == 0 {
+		result.PendingActions = pointer.To([]UpgradeTask{})
+		engines, err := e.kubeClient.ListDatabaseEngines(ctx, namespace)
+		if err != nil {
+			return err
+		}
+		for _, engine := range engines.Items {
+			check, err := e.checkDatabases(ctx, &engine)
+			if err != nil {
+				return err
+			}
+			for _, c := range check {
+				*result.PendingActions = append(*result.PendingActions, UpgradeTask{
+					Name:        c.Name,
+					PendingTask: pointer.To(UpgradeTaskPendingTask(pointer.Get(c.PendingTask))),
+					Message:     c.Message,
+				})
+			}
+		}
+	}
+	return c.JSON(http.StatusOK, result)
+}
+
+// ApproveUpgradePlan starts the upgrade of operators in the provided namespace.
+func (e *EverestServer) ApproveUpgradePlan(c echo.Context, namespace string) (err error) {
+	ctx := c.Request().Context()
+
+	up, err := e.getUpgradePlan(ctx, namespace)
+	if err != nil {
+		return err
+	}
+
+	// lock all engines that will be upgraded.
+	if err := e.setLockDBEnginesForUpgrade(ctx, namespace, up, true); err != nil {
+		return errors.Join(err, errors.New("failed to lock engines"))
+	}
+	// unlock if we return with error.
+	defer func() {
+		if err != nil {
```

Comment:
I don't think this condition will ever be true...
This `err` that you're checking here is the `err` returned from `up, err := e.getUpgradePlan(ctx, namespace)` in line 109.
The `err` in line 137 `if err := e.startOperatorUpgradeWithRetry(ctx, "", namespace, ""); err != nil {` only exist in the scope of the conditional and won't get picked up by the deferred function.

In order to get it in the correct scope you'll need to remove the `:` from the `:=` operator

Reply:
What about the return in line 131? Wouldn't we want to unlock there as well?

Reply:
> won't get picked up by the deferred function.

That's not true, it will get picked up by the defer function. Any value I return after the defer will be contained in `err`. You can test this with a simply program:

```go
package main

import "fmt"

func testDefer() (m int) {
	m = 0
	defer func() {
		if m == 3 {
			fmt.Println("value is 3")
		}
	}()
	return 3
}

func main() {
	testDefer()
}
```

Reply:
You're completely correct, I also tested this with a simple program just like yours but I was blind to my own test results because I got the exact same result as you and my mind saw something else... :face_exhaling: fuck confirmation bias :facepalm:

Reply:
Regardless of that...
> What about the return in line 131? Wouldn't we want to unlock there as well?

Reply:
Yeah, we need to unlock on 131. For some reason, I thought returning a non-200 status in `c.JSON` results in an error being returned, but that doesn't make sense.. will update

Reply:
Updated

I ended up removing the defer and named return, since I had to explicitly call unlock anyway on line 131. So the only place where the defer could have been triggered is towards the end of the function. Not worth the added complexity..